modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/RowHandler.java

```diff
@@ -31,6 +31,9 @@
     /** */
     Object get(int field, Row row);
 
+    /** */
+    Object getBiRows(int field, Object row1, Object row2);
```

Comment:
I think row handler is not the right place for such a method. Everything this method does can be done by some external utility class using `get()` and `columnCount()` methods.

Reply:
you are right but unfortunately this method can`t be static ... otherwise exception will be thrown :
java.lang.AssertionError
	at org.apache.calcite.linq4j.tree.MethodCallExpression.<init>(MethodCallExpression.java:43)
	at org.apache.calcite.linq4j.tree.MethodCallExpression.<init>(MethodCallExpression.java:53)
	at org.apache.calcite.linq4j.tree.Expressions.call(Expressions.java:377)
	at org.apache.ignite.internal.processors.query.calcite.exec.exp.ExpressionFactoryImpl$BiFieldGetter.fillExpressions(ExpressionFactoryImpl.java:519) 

this check : assert targetExpression == null == Modifier.isStatic(method.getModifiers());

Reply:
i hope i understand your way ) but seems it wrong, this implementation fully depends on  RowHandler implementation, look - if i implement additional RowHandler impl based on BO for copying preventing i will need the different method in such a case.

Reply:
got it, fixed.